pkg/cache/internal/informers_map_test.go

```diff
@@ -0,0 +1,122 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package internal
+
+import (
+	"testing"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+type testKind struct {
+	gvk schema.GroupVersionKind
+}
+
+func (o *testKind) SetGroupVersionKind(kind schema.GroupVersionKind) {
+	o.gvk = kind
+}
+
+func (o *testKind) GroupVersionKind() schema.GroupVersionKind {
+	return o.gvk
+}
+
+type testType struct {
+	kind  schema.ObjectKind
+	label string
+}
+
+func newTestType(label string) testType {
+	return testType{
+		kind: &testKind{
+			gvk: schema.GroupVersionKind{
+				Group:   "originalgroup",
+				Version: "OriginalVersion",
+				Kind:    "OriginalKind",
+			},
+		},
+		label: label,
+	}
+}
+
+func (obj testType) GetObjectKind() schema.ObjectKind { return obj.kind }
+func (obj testType) DeepCopyObject() runtime.Object   { return obj }
+
+// TestGVKFixupWatcher tests that gvkFixupWatcher behaves like the watch that it
+// wraps and that it overrides the GVK.
+// Adapted from https://github.com/kubernetes/kubernetes/blob/adbda068c1808fcc8a64a94269e0766b5c46ec41/staging/src/k8s.io/apimachinery/pkg/watch/watch_test.go#L33-L78
+func TestGVKFixupWatcher(t *testing.T) {
+	f := watch.NewFake()
+	// This is the GVK which we expect the wrapper to set on all the events
+	expectedGVK := schema.GroupVersionKind{
+		Group:   "testgroup",
+		Version: "v999",
+		Kind:    "TestKind",
+	}
+	gvkfw := newGVKFixupWatcher(expectedGVK, f)
+
+	table := []struct {
+		t watch.EventType
+		s testType
+	}{
+		{watch.Added, newTestType("foo")},
+		{watch.Modified, newTestType("qux")},
+		{watch.Modified, newTestType("bar")},
+		{watch.Deleted, newTestType("bar")},
+		{watch.Error, newTestType("error: blah")},
+	}
+
+	// Prove that f implements Interface by phrasing this as a function.
+	consumer := func(w watch.Interface) {
+		for _, expect := range table {
+			got, ok := <-w.ResultChan()
+			if !ok {
+				t.Fatalf("closed early")
+			}
+			if e, a := expect.t, got.Type; e != a {
+				t.Fatalf("Expected %v, got %v", e, a)
+			}
+			a, ok := got.Object.(testType)
+			if !ok {
+				t.Fatalf("Unexpected type: %T", got.Object)
+			}
+			if a.label != expect.s.label {
+				t.Fatalf("Expected %#v, got %#v", expect.s, a)
+			}
+			if actualGVK := got.Object.GetObjectKind().GroupVersionKind(); expectedGVK != actualGVK {
+				t.Fatalf("Expected %v, got %v", expectedGVK, actualGVK)
+			}
+		}
+		_, stillOpen := <-w.ResultChan()
+		if stillOpen {
+			t.Fatal("Never stopped")
+		}
+	}
+
+	sender := func() {
+		f.Add(newTestType("foo"))
+		f.Action(watch.Modified, newTestType("qux"))
+		f.Modify(newTestType("bar"))
+		f.Delete(newTestType("bar"))
+		f.Error(newTestType("error: blah"))
+		f.Stop()
```

Comment:
Note: The test sends events to the  wrapped watcher and then checks the results that come from the wrapper.